core/consensus/babe/src/lib.rs

```diff
@@ -75,28 +155,918 @@ impl Decode for BabeSeal {
+	pub force_authoring: bool,
+}
+
+/// Start the babe worker. The returned future should be run in a tokio runtime.
+pub fn start_babe<B, C, E, I, SO, Error, OnExit>(BabeParams {
+	config,
+	local_key,
+	client,
+	block_import,
+	env,
+	sync_oracle,
+	on_exit,
+	inherent_data_providers,
+	force_authoring,
+}: BabeParams<C, E, I, SO, OnExit>) -> Result<impl Future<Item=(), Error=()>, consensus_common::Error> where
+	B: Block,
+	C: ChainHead<B> + ProvideRuntimeApi + ProvideCache<B>,
+	C::Api: AuthoritiesApi<B>,
+	E: Environment<B, Error=Error>,
+	E::Proposer: Proposer<B, Error=Error>,
+	<<E::Proposer as Proposer<B>>::Create as IntoFuture>::Future: Send + 'static,
+	I: BlockImport<B> + Send + Sync + 'static,
+	SO: SyncOracle + Send + Sync + Clone,
+	DigestItemFor<B>: CompatibleDigestItem + DigestItem<AuthorityId=Public>,
+	Error: ::std::error::Error + Send + From<::consensus_common::Error> + From<I::Error> + 'static,
+	OnExit: Future<Item=(), Error=()>,
+{
+	let worker = BabeWorker {
+		client: client.clone(),
+		block_import,
+		env,
+		local_key,
+		inherent_data_providers: inherent_data_providers.clone(),
+		sync_oracle: sync_oracle.clone(),
+		force_authoring,
+		threshold: config.threshold(),
+	};
+	slots::start_slot_worker::<_, _, _, _, _, BabeSlotCompatible, _>(
+		config.0,
+		client,
+		Arc::new(worker),
+		sync_oracle,
+		on_exit,
+		inherent_data_providers
+	)
+}
+
+struct BabeWorker<C, E, I, SO> {
+	client: Arc<C>,
+	block_import: Arc<I>,
+	env: Arc<E>,
+	local_key: Arc<sr25519::Pair>,
+	sync_oracle: SO,
+	inherent_data_providers: InherentDataProviders,
+	force_authoring: bool,
+	threshold: u64,
+}
+
+impl<B: Block, C, E, I, Error, SO> SlotWorker<B> for BabeWorker<C, E, I, SO> where
+	C: ProvideRuntimeApi + ProvideCache<B>,
+	C::Api: AuthoritiesApi<B>,
+	E: Environment<B, Error=Error>,
+	E::Proposer: Proposer<B, Error=Error>,
+	<<E::Proposer as Proposer<B>>::Create as IntoFuture>::Future: Send + 'static,
+	I: BlockImport<B> + Send + Sync + 'static,
+	SO: SyncOracle + Send + Clone,
+	DigestItemFor<B>: CompatibleDigestItem + DigestItem<AuthorityId=Public>,
+	Error: std::error::Error + Send + From<::consensus_common::Error> + From<I::Error> + 'static,
+{
+	type OnSlot = Box<Future<Item=(), Error=consensus_common::Error> + Send>;
+
+	fn on_start(
+		&self,
+		slot_duration: u64
+	) -> Result<(), consensus_common::Error> {
+		register_babe_inherent_data_provider(&self.inherent_data_providers, slot_duration)
+	}
+
+	fn on_slot(
+		&self,
+		chain_head: B::Header,
+		slot_info: SlotInfo,
+	) -> Self::OnSlot {
+		let pair = self.local_key.clone();
+		let _public_key = self.local_key.public();
+		let client = self.client.clone();
+		let block_import = self.block_import.clone();
+		let env = self.env.clone();
+
+		let (timestamp, slot_num, slot_duration) =
+			(slot_info.timestamp, slot_info.number, slot_info.duration);
+
+		let authorities = match authorities(client.as_ref(), &BlockId::Hash(chain_head.hash())) {
```

Comment:
Probably the `Vec<_>` could be turned into a `HashSet<_>` of authorities here so the code doesn't have the `O(n)` problem further on. However of the `.contains` is only run once during the lifetime of this set, it might be even slower.

Reply:
Correct. We only run `.contains` once.